src/image/ImageBaseFitsReader.cc

```diff
@@ -0,0 +1,209 @@
+}
+
+namespace {
+
+void checkFitsFile(fits::Fits * f) {
+    if (f == nullptr) {
+        throw LSST_EXCEPT(
+            fits::FitsError,
+            "FitsReader not initialized; desired HDU is probably missing."
+        );
+    }
+}
+
+} // anonymous
+
+std::string ImageBaseFitsReader::readDType() const {
+    checkFitsFile(_fitsFile);
+    fits::HduMoveGuard guard(*_fitsFile, _hdu);
+    return _fitsFile->getImageDType();
+}
+
+lsst::geom::Box2I ImageBaseFitsReader::readBBox(ImageOrigin origin) {
+    readMetadata();  // guarantees _bbox is initialized
+    if (origin == LOCAL) {
+        return lsst::geom::Box2I(lsst::geom::Point2I(), _bbox.getDimensions());
+    }
+    return _bbox;
+}
+
+lsst::geom::Point2I ImageBaseFitsReader::readXY0(lsst::geom::Box2I const & bbox, ImageOrigin origin) {
+    if (bbox.isEmpty()) {
+        return readBBox().getMin();
+    } else if (origin == PARENT) {
+        return bbox.getMin();
+    } else {
+        auto full = readBBox();
+        return full.getMin() + lsst::geom::Extent2I(bbox.getMin());
+    }
+}
+
+std::shared_ptr<daf::base::PropertyList> ImageBaseFitsReader::readMetadata() {
+    checkFitsFile(_fitsFile);
+    if (_metadata == nullptr) {
+        fits::HduMoveGuard guard(*_fitsFile, _hdu);
+        auto metadata = fits::readMetadata(*_fitsFile, /*strip=*/true);
+        // Construct _bbox and check dimensions at the same time both to make
+        // one less valid state for this class to be in and to make sure
+        // the appropriate metadata is stripped before we ever return it.
+        int nAxis = _fitsFile->getImageDim();
+        ndarray::Vector<ndarray::Size, 2> shape;
+        if (nAxis == 2) {
+            shape = _fitsFile->getImageShape<2>();
+        } else if (nAxis == 3) {
+            ndarray::Vector<ndarray::Size, 3> shape3 = _fitsFile->getImageShape<3>();
+            if (shape3[0] != 1) {
+                throw LSST_EXCEPT(
+                    fits::FitsError,
+                    str(boost::format("Error reading %s: HDU %d has 3rd dimension %d != 1") %
+                        getFileName() % _hdu % shape3[0]));
+            }
+            shape = shape3.last<2>();
+        } else {
+            throw LSST_EXCEPT(
+                fits::FitsError,
+                str(boost::format("Error reading %s: HDU %d has %d dimensions") %
+                    getFileName() % _hdu % nAxis));
+        }
+        auto xy0 = afw::geom::getImageXY0FromMetadata(*metadata, detail::wcsNameForXY0, /*strip=*/true);
+        _bbox = lsst::geom::Box2I(xy0, lsst::geom::Extent2I(shape[1], shape[0]));
+        _metadata = std::move(metadata);
+    }
+    return _metadata;
+}
+
+namespace {
+
+// Return a numpy-like description of a primitive type (e.g. uint8, float64).
+template <typename T>
+std::string makeDTypeString() {
+    using L = std::numeric_limits<T>;
+    static_assert(L::is_specialized, "makeDTypeString requires a primitive numeric template parameter.");
+    std::string result;
```

Comment:
I don't think the lambda-if-else trick works in this case because these clauses are appending the result, not assigning something totally new to it.  And I don't think `switch` is appropriate because we only have one two possibilities at each level (i.e. there are no `else if` clauses).